File: typescript/examples/04_audit_trust_chain.ts
/**
 * Parmana TypeScript SDK
 *
 * Example 04
 *
 * Audit Execution Trust Record
 */

import {
  ExecutionTrustRecord,
} from "../src/index.js";

function auditTrustRecord(
  record: ExecutionTrustRecord,
): void {
  console.log("======================================");
  console.log(" Parmana TypeScript SDK");
  console.log(" Example 04 - Audit Trust Chain");
  console.log("======================================");
  console.log();

  console.log("Execution Trust Record");
  console.log("----------------------");
  console.log(
    "Trust Record ID :",
    record.trustRecordId,
  );
  console.log(
    "Transaction ID  :",
    record.businessTransactionId,
  );
  console.log(
    "Trust Hash      :",
    record.trustRecordHash,
  );
  console.log(
    "Created         :",
    record.createdAt,
  );
  console.log(
    "Updated         :",
    record.updatedAt,
  );

  console.log();

  console.log("Business Transaction");
  console.log("----------------------");
  console.log(
    "Authority        :",
    record.transaction.authority.authorityName,
  );
  console.log(
    "Authorization ID :",
    record.transaction.authorization.authorizationId,
  );
  console.log(
    "Intent           :",
    record.transaction.intent.operation,
  );
  console.log(
    "Policy           :",
    `${record.transaction.policy.policyName} (${record.transaction.policy.policyVersion})`,
  );

  console.log();

  console.log("Execution Summary");
  console.log("----------------------");
  console.log(
    "Executions    :",
    record.executions.length,
  );
  console.log(
    "Overrides     :",
    record.overrides.length,
  );
  console.log(
    "Verifications :",
    record.verifications.length,
  );
  console.log(
    "Receipts      :",
    record.receipts.length,
  );

  console.log();

  console.log("Execution History");
  console.log("----------------------");

  if (record.executions.length === 0) {
    console.log("No executions recorded.");
  }

  for (const execution of record.executions) {
    console.log();
    console.log(
      "Execution ID :",
      execution.executionId,
    );
    console.log(
      "Status       :",
      execution.status,
    );
    console.log(
      "Mode         :",
      execution.mode,
    );
    console.log(
      "Decision     :",
      execution.decision.outcome,
    );
    console.log(
      "Started      :",
      execution.startedAt,
    );
    console.log(
      "Completed    :",
      execution.completedAt ?? "-",
    );
  }

  console.log();

  console.log("Verification History");
  console.log("----------------------");

  if (record.verifications.length === 0) {
    console.log("No verifications recorded.");
  }

  for (const verification of record.verifications) {
    console.log();
    console.log(
      "Verification ID :",
      verification.verificationId,
    );
    console.log(
      "Status          :",
      verification.status,
    );
    console.log(
      "Verified At     :",
      verification.verifiedAt,
    );
  }

  console.log();

  console.log("Receipt History");
  console.log("----------------------");

  if (record.receipts.length === 0) {
    console.log("No receipts recorded.");
  }

  for (const receipt of record.receipts) {
    console.log();
    console.log(
      "Receipt ID :",
      receipt.receiptId,
    );
    console.log(
      "Algorithm  :",
      receipt.algorithm,
    );
    console.log(
      "Issued At  :",
      receipt.issuedAt,
    );
  }

  console.log();

  console.log("Audit Result");
  console.log("----------------------");
  console.log(
    "✓ Authority recorded",
  );
  console.log(
    "✓ Authorization recorded",
  );
  console.log(
    "✓ Intent recorded",
  );
  console.log(
    "✓ Policy Reference recorded",
  );
  console.log(
    "✓ Execution history available",
  );
  console.log(
    "✓ Verification history available",
  );
  console.log(
    "✓ Receipt history available",
  );
  console.log(
    "✓ Execution Trust Record is suitable for independent audit",
  );
}

async function main(): Promise<void> {
  /**
   * Placeholder Trust Record.
   *
   * In production this would be retrieved from the
   * Parmana Runtime or Storage service.
   */
  const trustRecord = {
    trustRecordId: "trust-record-001",

    businessTransactionId: "txn-001",

    transaction: {} as never,

    overrides: [],

    executions: [],

    verifications: [],

    receipts: [],

    trustRecordHash:
      "7f84e2d3c1d9d8d3d2c8f7e9b6a4c5d2",

    createdAt: new Date(),

    updatedAt: new Date(),
  } satisfies ExecutionTrustRecord;

  auditTrustRecord(trustRecord);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: python-old/examples/04_audit_trust_chain.py
"""
Parmana Python SDK

Example 04

Audit Trust Chain

Demonstrates how an auditor can inspect the
complete Execution Trust Record.

Authority
        ↓
Authorization
        ↓
Intent
        ↓
Policy
        ↓
Decision
        ↓
Execution
        ↓
Verification
        ↓
Receipt
"""

from datetime import datetime
from uuid import uuid4

from parmana.models.authority import Authority
from parmana.models.authorization import Authorization
from parmana.models.business_transaction import BusinessTransaction
from parmana.models.decision import (
    Decision,
    DecisionOutcome,
)
from parmana.models.execution import (
    Execution,
    ExecutionMode,
    ExecutionStatus,
)
from parmana.models.execution_trust_record import (
    ExecutionTrustRecord,
)
from parmana.models.intent import Intent
from parmana.models.policy_reference import PolicyReference
from parmana.models.receipt import Receipt
from parmana.models.verification import (
    Verification,
    VerificationStatus,
)


def audit(record: ExecutionTrustRecord) -> None:
    """
    Placeholder audit implementation.
    """

    print()
    print("========== TRUST CHAIN ==========")

    print(f"Authority           : {record.transaction.authority.authority_name}")
    print(f"Authorization       : {record.transaction.authorization.authorization_id}")
    print(f"Intent              : {record.transaction.intent.operation}")

    print(
        f"Policy              : "
        f"{record.transaction.policy_reference.policy_name} "
        f"v{record.transaction.policy_reference.policy_version}"
    )

    print()

    print(f"Executions          : {len(record.executions)}")
    print(f"Verifications       : {len(record.verifications)}")
    print(f"Receipts            : {len(record.receipts)}")

    print()

    print(f"Trust Record Hash   : {record.trust_record_hash}")

    print()

    print("Trust chain audit completed successfully.")


def main() -> None:

    authority = Authority(
        authority_id=str(uuid4()),
        authority_type="ORGANIZATION",
        authority_name="Acme Corporation",
        created_at=datetime.utcnow(),
        public_key="public-key",
        signature_algorithm="Ed25519",
    )

    authorization = Authorization(
        authorization_id=str(uuid4()),
        authority_id=authority.authority_id,
        subject="robot-01",
        action="MOVE",
        resource="warehouse",
        issued_at=datetime.utcnow(),
    )

    intent = Intent(
        intent_id=str(uuid4()),
        operation="MOVE",
        target="warehouse-zone-a",
        parameters={},
    )

    policy = PolicyReference(
        policy_name="warehouse-policy",
        policy_version="1.0.0",
    )

    transaction = BusinessTransaction(
        business_transaction_id=str(uuid4()),
        authority=authority,
        authorization=authorization,
        intent=intent,
        policy_reference=policy,
        created_at=datetime.utcnow(),
    )

    decision = Decision(
        decision_id=str(uuid4()),
        intent_id=intent.intent_id,
        policy=policy,
        signals={},
        outcome=DecisionOutcome.APPROVED,
        evaluated_at=datetime.utcnow(),
    )

    execution = Execution(
        execution_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        decision=decision,
        status=ExecutionStatus.COMPLETED,
        mode=ExecutionMode.SYNC,
        started_at=datetime.utcnow(),
        completed_at=datetime.utcnow(),
        evidence={},
    )

    verification = Verification(
        verification_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        status=VerificationStatus.VERIFIED,
        verified_at=datetime.utcnow(),
        trust_record_hash="abc123",
    )

    receipt = Receipt(
        receipt_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        execution_id=execution.execution_id,
        trust_record_hash="abc123",
        receipt_hash="receipt-hash",
        signature="signature",
        algorithm="Ed25519",
        issued_at=datetime.utcnow(),
    )

    record = ExecutionTrustRecord(
        trust_record_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        transaction=transaction,
        overrides=(),
        executions=(execution,),
        verifications=(verification,),
        receipts=(receipt,),
        trust_record_hash="abc123",
        created_at=datetime.utcnow(),
        updated_at=datetime.utcnow(),
    )

    print()
    print("===================================")
    print("Parmana Trust Chain Audit")
    print("===================================")

    audit(record)

    print("===================================")


if __name__ == "__main__":
    main()